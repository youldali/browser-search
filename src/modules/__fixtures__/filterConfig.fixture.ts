import { 
    buildFilterConfigData, 
    Filter, 
    FilterConfig,
    FiltersByGroup,
    FilterToGroup,
} from '../filterConfiguration/filterConfig.model';

import { Operators } from '../filterConfiguration/operators';

export const filterDictionary: Dictionary<Filter> = {
    priceMin: { id: 'priceMin', field: 'price', operator: Operators.gt, operand: 200 },
    priceMax: { id: 'priceMax', field: 'price', operator: Operators.lt, operand: 500 },
    numberOfPeople: { id: 'numberOfPeople', field: 'numberOfPeople', operator: Operators.equals, operand: 2 },
    "activity-1": { id: 'activity-1', field: 'activity', operator: Operators.contains, operand: 'swimming' },
    "activity-2": { id: 'activity-2', field: 'activity', operator: Operators.contains, operand: 'tennis' },
    "activity-3": { id: 'activity-3', field: 'activity', operator: Operators.contains, operand: 'golfing' },
};

export const filterByGroup: FiltersByGroup = {
    '0': [filterDictionary.priceMin],
    '1': [filterDictionary.priceMax],
    '2': [filterDictionary.numberOfPeople],
    '3': [filterDictionary['activity-1'], filterDictionary['activity-2'], filterDictionary['activity-3']],
}

export const filterToGroup: FilterToGroup = new Map([
    [filterDictionary.priceMin.id, '0'],
    [filterDictionary.priceMax.id, '1'],
    [filterDictionary.numberOfPeople.id, '2'],
    [filterDictionary['activity-1'].id, '3'],
    [filterDictionary['activity-2'].id, '3'],
    [filterDictionary['activity-3'].id, '3'],
]);

export const filterConfig: FilterConfig = [
    [filterDictionary.priceMin],
    [filterDictionary.priceMax],
    [filterDictionary.numberOfPeople],
    [filterDictionary['activity-1'], filterDictionary['activity-2'], filterDictionary['activity-3']]
];

export const filtersApplied = [filterDictionary.priceMin, filterDictionary['activity-1'], filterDictionary['activity-2']];
export const filtersIdsApplied = ['priceMin', 'activity-1', 'activity-2'];
export const filterConfigData = buildFilterConfigData(filterConfig)(filtersIdsApplied);
