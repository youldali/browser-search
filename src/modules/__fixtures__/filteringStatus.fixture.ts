import { getFilterStatusForItem, FilteredBoxStatus } from '../filtering';
import { filteringData } from './filtering.fixture';

export const items: any[] = [
    { id: 0, price: 250, numberOfPeople: 1, activity: ['swimming', 'tennis']},
    { id: 1, price: 1000, numberOfPeople: 2, activity: ['football', 'swimming', 'tennis', 'golfing']},
    { id: 2, price: 20, numberOfPeople: 5, activity: ['swimming', 'tennis'] },
    { id: 3, price: 1000, numberOfPeople: 1, activity: ['football', 'golfing']},
    { id: 4, price: 10, numberOfPeople: 2, activity: ['football']},
];

export const itemToFilteringStatus:  Map<any, FilteredBoxStatus> = new Map (
    items.map( item => [
        item,
        getFilterStatusForItem(filteringData)(item),
    ])
);
