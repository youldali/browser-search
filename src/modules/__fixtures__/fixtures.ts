import { 
    buildFilterConfigData, 
    Filter, 
    FilterConfig,
    FiltersByGroup,
} from '../filterConfiguration/filterConfig.model';
import { Operators, operatorToFunction } from '../filterConfiguration/operators';
import { 
    FilterFunctionsCollection,
    FilterGroupToFilterFunctions,
    FilterFunctionsToFilterGroup,
    FilteredItemStatus,
    getFilteringFunctionsData,
    getFilterStatusForItem,
} from '../filteringStatus';
import { createFilteringData, FilterIdToMatchingItemIds } from '../filteringData';


export const items: any[] = [
    { id: 0, price: 250, numberOfPeople: 1, activity: ['swimming', 'tennis']},
    { id: 1, price: 1000, numberOfPeople: 2, activity: ['football', 'swimming', 'tennis', 'golfing']},
    { id: 2, price: 20, numberOfPeople: 5, activity: ['swimming', 'tennis'] },
    { id: 3, price: 1000, numberOfPeople: 1, activity: ['football', 'golfing']},
    { id: 4, price: 10, numberOfPeople: 2, activity: ['football']},
];

/**
 * FilterConfiguration Module
 */
export const filterDictionary: Dictionary<Filter> = {
    priceMin: { id: 'priceMin', field: 'price', operator: Operators.gt, operand: 200 },
    priceMax: { id: 'priceMax', field: 'price', operator: Operators.lt, operand: 500 },
    numberOfPeople: { id: 'numberOfPeople', field: 'numberOfPeople', operator: Operators.equals, operand: 2 },
    "activity-1": { id: 'activity-1', field: 'activity', operator: Operators.contains, operand: 'swimming' },
    "activity-2": { id: 'activity-2', field: 'activity', operator: Operators.contains, operand: 'tennis' },
    "activity-3": { id: 'activity-3', field: 'activity', operator: Operators.contains, operand: 'golfing' },
};

export const filterByGroup: FiltersByGroup = {
    '0': [filterDictionary.priceMin],
    '1': [filterDictionary.priceMax],
    '2': [filterDictionary.numberOfPeople],
    '3': [filterDictionary['activity-1'], filterDictionary['activity-2'], filterDictionary['activity-3']],
}

export const filterConfig: FilterConfig = [
    [filterDictionary.priceMin],
    [filterDictionary.priceMax],
    [filterDictionary.numberOfPeople],
    [filterDictionary['activity-1'], filterDictionary['activity-2'], filterDictionary['activity-3']]
];

export const filtersIdsApplied = ['priceMin', 'activity-1', 'activity-2'];
export const filterConfigData = buildFilterConfigData(filterConfig)(filtersIdsApplied);

/**
 * Filtering Module
 */

const priceMinFilterFunction = (target: any) => 
    operatorToFunction[filterDictionary.priceMin.operator](target?.[filterDictionary.priceMin.field], filterDictionary.priceMin.operand);

const activity1FilterFunction = (target: any) => 
    operatorToFunction[filterDictionary['activity-1'].operator](target?.[filterDictionary['activity-1'].field], filterDictionary['activity-1'].operand);

const activity2FilterFunction = (target: any) => 
    operatorToFunction[filterDictionary['activity-2'].operator](target?.[filterDictionary['activity-2'].field], filterDictionary['activity-2'].operand);

export const filterFunctionsCollection: FilterFunctionsCollection = [
    [priceMinFilterFunction],
    [activity1FilterFunction, activity2FilterFunction]
];

export const filterGroupToFilterFunctions: FilterGroupToFilterFunctions = {
    '0': filterFunctionsCollection[0],
    '3': filterFunctionsCollection[1],
}

export const filterFunctionsToFilterGroup: FilterFunctionsToFilterGroup = new Map([
    [filterGroupToFilterFunctions['0'], '0'],
    [filterGroupToFilterFunctions['3'], '3'],
]);

export const filteringData = getFilteringFunctionsData(filterConfigData);

export const itemToFilteringStatus:  Map<any, FilteredItemStatus> = new Map (
    items.map( item => [
        item,
        getFilterStatusForItem(filteringData)(item),
    ])
);


/**
 * FilterStatistics Module
 */

export const filterIdToMatchingItemIds: FilterIdToMatchingItemIds = {
    [filterDictionary.priceMin.id]: [0, 1, 3],
    [filterDictionary.priceMax.id]: [0, 2, 4],
    [filterDictionary.numberOfPeople.id]: [1, 4],
    [filterDictionary['activity-1'].id]: [0, 1, 2],
    [filterDictionary['activity-2'].id]: [0, 1, 2],
    [filterDictionary['activity-3'].id]: [1, 3],
}

const filteringStatisticsDataBuilder = createFilteringData(filterConfigData)(filterIdToMatchingItemIds);
itemToFilteringStatus.forEach( (filteredItemStatus, item) => {
        filteringStatisticsDataBuilder.addFilteredObjectStatus(filteredItemStatus, item.id)
    }
);
export const filteringStatisticsData = filteringStatisticsDataBuilder.done();