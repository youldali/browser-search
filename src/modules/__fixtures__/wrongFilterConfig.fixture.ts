
export const emptyFilterConfig: any = [
];

export const emptyGroupOfFilters: any = [
    [ 
        { id: 'priceMin', field: 'price', operator: 'gt', operand: 200 }, 
    ],
    [ 
    ]
];

export const incompleteFilter: any = [
    [ 
        { field: 'price', operator: 'gt', operand: 200 }, 
        { id: 'priceMax', field: 'price', operator: 'lt', operand: 500 }, 
    ],
];

export const invalidOperator: any = [
    [ 
        { field: 'price', operator: '.....', operand: 200 }, 
    ],
];