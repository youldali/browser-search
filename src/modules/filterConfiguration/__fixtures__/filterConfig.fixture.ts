import { 
    buildFilterConfigData, 
    Filter, 
    FilterConfig,
    FiltersByGroup,
    FilterToGroup,
} from '../filterConfig.model';

import { 
    FilterFunctionsCollection,
    FilterGroupToFilterFunctions,
    FilterFunctionsToFilterGroup,
} from '../../filtering/filter.model';

import { Operators, operatorToFunction } from '../operators';

export const filterDictionary: Dictionary<Filter> = {
    priceMin: { id: 'priceMin', field: 'price', operator: Operators.gt, operand: 200 },
    priceMax: { id: 'priceMax', field: 'price', operator: Operators.lt, operand: 500 },
    numberOfPeople: { id: 'numberOfPeople', field: 'numberOfPeople', operator: Operators.equals, operand: 2 },
    "activity-1": { id: 'activity-1', field: 'activity', operator: Operators.contains, operand: 'swimming' },
    "activity-2": { id: 'activity-2', field: 'activity', operator: Operators.contains, operand: 'tennis' },
    "activity-3": { id: 'activity-3', field: 'activity', operator: Operators.contains, operand: 'golfing' },
};

export const filterByGroup: FiltersByGroup = {
    '0': [filterDictionary.priceMin, filterDictionary.priceMax],
    '1': [filterDictionary.numberOfPeople],
    '2': [filterDictionary['activity-1'], filterDictionary['activity-2'], filterDictionary['activity-3']],
}

export const filterToGroup: FilterToGroup = new Map([
    [filterDictionary.priceMin, '0'],
    [filterDictionary.priceMax, '0'],
    [filterDictionary.numberOfPeople, '1'],
    [filterDictionary['activity-1'], '2'],
    [filterDictionary['activity-2'], '2'],
    [filterDictionary['activity-3'], '2'],
]);

export const filterConfig: FilterConfig = [
    [filterDictionary.priceMin, filterDictionary.priceMax],
    [filterDictionary.numberOfPeople],
    [filterDictionary['activity-1'], filterDictionary['activity-2'], filterDictionary['activity-3']]
];

export const filtersApplied = [filterDictionary.priceMin, filterDictionary['activity-1'], filterDictionary['activity-2']];

export const filtersIdsApplied = ['priceMin', 'activity-1', 'activity-2'];

export const filterConfigData = buildFilterConfigData(filterConfig)(filtersIdsApplied);

export const priceMinFilterFunction = (target: any) => 
    operatorToFunction[filterDictionary.priceMin.operator](target?.[filterDictionary.priceMin.field], filterDictionary.priceMin.operand);

export const activity1FilterFunction = (target: any) => 
    operatorToFunction[filterDictionary['activity-1'].operator](target?.[filterDictionary['activity-1'].field], filterDictionary['activity-1'].operand);

export const activity2FilterFunction = (target: any) => 
    operatorToFunction[filterDictionary['activity-2'].operator](target?.[filterDictionary['activity-2'].field], filterDictionary['activity-2'].operand);

export const filterFunctionsCollection: FilterFunctionsCollection = [
    [priceMinFilterFunction],
    [activity1FilterFunction, activity2FilterFunction]
];

export const filterGroupToFilterFunctions: FilterGroupToFilterFunctions = {
    '0': filterFunctionsCollection[0],
    '2': filterFunctionsCollection[1],
}

export const filterFunctionsToFilterGroup: FilterFunctionsToFilterGroup = new Map([
    [filterGroupToFilterFunctions['0'], '0'],
    [filterGroupToFilterFunctions['2'], '2'],
]);

export const sampleData = [
    { price: 50, numberOfPeople: 1, activity: ['swimming'] },
    { price: 550, numberOfPeople: 2, activity: ['swimming', 'tennis'] },
    { price: 200, numberOfPeople: 1, activity: [] },
]