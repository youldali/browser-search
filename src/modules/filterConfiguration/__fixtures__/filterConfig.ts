import { FilterConfig } from '../filterConfig.model';
import { Operators } from '../operators'

export const validFilterConfig: FilterConfig = [
    [ 
        { id: 'priceMin', field: 'price', operator: Operators.gt, operand: 200 }, 
        { id: 'priceMax', field: 'price', operator: Operators.lt, operand: 500 }, 
    ],
    [ 
        { id: 'numberOfPeople', field: 'numberOfPeople', operator: Operators.equals, operand: 2 }, 
    ],
    [ 
        { id: 'activity-1', field: 'activity', operator: Operators.contains, operand: 'swimming' }, 
        { id: 'activity-2', field: 'activity', operator: Operators.contains, operand: 'tennis' }, 
        { id: 'activity-3', field: 'activity', operator: Operators.contains, operand: 'golfing' }, 
    ]
];

export const emptyFilterConfig: any = [
];

export const emptyGroupOfFilters: any = [
    [ 
        { id: 'priceMin', field: 'price', operator: Operators.gt, operand: 200 }, 
    ],
    [ 
    ]
];

export const incompleteFilter: any = [
    [ 
        { field: 'price', operator: Operators.gt, operand: 200 }, 
        { id: 'priceMax', field: 'price', operator: Operators.lt, operand: 500 }, 
    ],
];

export const invalidOperator: any = [
    [ 
        { field: 'price', operator: '.....', operand: 200 }, 
    ],
];