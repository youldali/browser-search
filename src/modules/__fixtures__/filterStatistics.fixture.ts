import { filterConfigData, filterDictionary } from './filterConfig.fixture';
import { itemToFilteringStatus } from './filteringStatus.fixture';
import { createFilteringStatistics, FilterIdToMatchingItemIds } from '../filteringStatistics';


export const filterIdToMatchingItemIds: FilterIdToMatchingItemIds = {
    [filterDictionary.priceMin.id]: [0, 1, 3],
    [filterDictionary.priceMax.id]: [0, 2, 4],
    [filterDictionary.numberOfPeople.id]: [1, 4],
    [filterDictionary['activity-1'].id]: [0, 1, 2],
    [filterDictionary['activity-2'].id]: [0, 1, 2],
    [filterDictionary['activity-3'].id]: [1, 3],
}

const filteringStatisticsDataBuilder = createFilteringStatistics(filterConfigData)(filterIdToMatchingItemIds);
itemToFilteringStatus.forEach( (filteredItemStatus, item) => {
        filteringStatisticsDataBuilder.addFilteredObjectStatus(filteredItemStatus, item.id)
    }
);
export const filteringStatisticsData = filteringStatisticsDataBuilder.done();
