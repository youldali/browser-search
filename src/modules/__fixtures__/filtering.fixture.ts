import { filterConfigData, filterDictionary } from './filterConfig.fixture'
import { operatorToFunction } from '../filterConfiguration/operators';
import { 
    FilterFunctionsCollection,
    FilterGroupToFilterFunctions,
    FilterFunctionsToFilterGroup,
    getFilteringData,
} from '../filtering/filtering.model';

const priceMinFilterFunction = (target: any) => 
    operatorToFunction[filterDictionary.priceMin.operator](target?.[filterDictionary.priceMin.field], filterDictionary.priceMin.operand);

const activity1FilterFunction = (target: any) => 
    operatorToFunction[filterDictionary['activity-1'].operator](target?.[filterDictionary['activity-1'].field], filterDictionary['activity-1'].operand);

const activity2FilterFunction = (target: any) => 
    operatorToFunction[filterDictionary['activity-2'].operator](target?.[filterDictionary['activity-2'].field], filterDictionary['activity-2'].operand);

export const filterFunctionsCollection: FilterFunctionsCollection = [
    [priceMinFilterFunction],
    [activity1FilterFunction, activity2FilterFunction]
];

export const filterGroupToFilterFunctions: FilterGroupToFilterFunctions = {
    '0': filterFunctionsCollection[0],
    '3': filterFunctionsCollection[1],
}

export const filterFunctionsToFilterGroup: FilterFunctionsToFilterGroup = new Map([
    [filterGroupToFilterFunctions['0'], '0'],
    [filterGroupToFilterFunctions['3'], '3'],
]);

export const sampleData = [
    { price: 50, numberOfPeople: 1, activity: ['swimming'] },
    { price: 550, numberOfPeople: 2, activity: ['swimming', 'tennis'] },
    { price: 200, numberOfPeople: 1, activity: [] },
];

export const filteringData = getFilteringData(filterConfigData);