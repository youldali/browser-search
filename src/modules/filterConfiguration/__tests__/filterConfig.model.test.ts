import { buildFilterConfigData } from '../filterConfig.model';
import { validFilterConfig } from '../__fixtures__/filterConfig'

describe('buildFilterConfigData', function(){
    const filtersApplied = ['priceMin', 'activity-1', 'activity-2'];
    const filterConfigData = buildFilterConfigData(validFilterConfig)(filtersApplied);
    
	test('it should return a dictionary of all filters', () => {
		expect(filterConfigData.getFilterDictionary()).toMatchSnapshot();
    });

    test('it should return only the filters applied', () => {
		expect(filterConfigData.getFiltersApplied()).toMatchSnapshot();
    });

    test('it should return a dictionary of groupOfFilter by key of group', () => {
        expect(filterConfigData.getFiltersByGroup()).toMatchSnapshot();
    });
});