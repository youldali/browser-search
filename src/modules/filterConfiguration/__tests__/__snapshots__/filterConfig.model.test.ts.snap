// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildFilterConfig it should return a dictionary of all filters 1`] = `
Object {
  "activity-1": Object {
    "field": "activity",
    "id": "activity-1",
    "operand": "swimming",
    "operator": "contains",
  },
  "activity-2": Object {
    "field": "activity",
    "id": "activity-2",
    "operand": "tennis",
    "operator": "contains",
  },
  "activity-3": Object {
    "field": "activity",
    "id": "activity-3",
    "operand": "golfing",
    "operator": "contains",
  },
  "numberOfPeople": Object {
    "field": "numberOfPeople",
    "id": "numberOfPeople",
    "operand": 2,
    "operator": "==",
  },
  "priceMax": Object {
    "field": "price",
    "id": "priceMax",
    "operand": 500,
    "operator": "<",
  },
  "priceMin": Object {
    "field": "price",
    "id": "priceMin",
    "operand": 200,
    "operator": ">",
  },
}
`;

exports[`buildFilterConfig it should return a dictionary of groupOfFilter by key of group 1`] = `
Object {
  "0": Array [
    Object {
      "field": "price",
      "id": "priceMin",
      "operand": 200,
      "operator": ">",
    },
    Object {
      "field": "price",
      "id": "priceMax",
      "operand": 500,
      "operator": "<",
    },
  ],
  "1": Array [
    Object {
      "field": "numberOfPeople",
      "id": "numberOfPeople",
      "operand": 2,
      "operator": "==",
    },
  ],
  "2": Array [
    Object {
      "field": "activity",
      "id": "activity-1",
      "operand": "swimming",
      "operator": "contains",
    },
    Object {
      "field": "activity",
      "id": "activity-2",
      "operand": "tennis",
      "operator": "contains",
    },
    Object {
      "field": "activity",
      "id": "activity-3",
      "operand": "golfing",
      "operator": "contains",
    },
  ],
}
`;

exports[`buildFilterConfig it should return only the filters applied 1`] = `
Array [
  Object {
    "field": "price",
    "id": "priceMin",
    "operand": 200,
    "operator": ">",
  },
  Object {
    "field": "activity",
    "id": "activity-1",
    "operand": "swimming",
    "operator": "contains",
  },
  Object {
    "field": "activity",
    "id": "activity-2",
    "operand": "tennis",
    "operator": "contains",
  },
]
`;
